pkg/tuf/keys.go

```diff
@@ -0,0 +1,165 @@
+func readKey(role data.RoleName, keyFilename string, retriever notary.PassRetriever) (data.PrivateKey, error) {
+	pemBytes, err := ioutil.ReadFile(keyFilename)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading input root key file: %v", err)
+	}
+	isEncrypted := true
+	if err = cryptoservice.CheckRootKeyIsEncrypted(pemBytes); err != nil {
+		if role == data.CanonicalRootRole {
+			return nil, err
+		}
+		isEncrypted = false
+	}
+	var privKey data.PrivateKey
+	if isEncrypted {
+		privKey, _, err = trustmanager.GetPasswdDecryptBytes(retriever, pemBytes, "", data.CanonicalRootRole.String())
+	} else {
+		privKey, err = utils.ParsePEMPrivateKey(pemBytes, "")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return privKey, nil
+}
+
+// importRootKey imports the root key from path then adds the key to repo
+// returns key ids
+// https://github.com/theupdateframework/notary/blob/f255ae779066dc28ae4aee196061e58bb38a2b49/cmd/notary/tuf.go#L413
+func importRootKey(rootKey string, nRepo client.Repository, retriever notary.PassRetriever) ([]string, error) {
+	var rootKeyList []string
+
+	if rootKey != "" {
+		privKey, err := readKey(data.CanonicalRootRole, rootKey, retriever)
+		if err != nil {
+			return nil, err
+		}
+		// add root key to repo
+		err = nRepo.GetCryptoService().AddKey(data.CanonicalRootRole, "", privKey)
+		if err != nil {
+			return nil, fmt.Errorf("Error importing key: %v", err)
+		}
+		rootKeyList = []string{privKey.ID()}
+	} else {
+		rootKeyList = nRepo.GetCryptoService().ListKeys(data.CanonicalRootRole)
+	}
+
+	if len(rootKeyList) > 0 {
+		// Chooses the first root key available, which is initialization specific
+		// but should return the HW one first.
+		rootKeyID := rootKeyList[0]
+		log.Debugf("Signy found root key, using: %s\n", rootKeyID)
+
+		return []string{rootKeyID}, nil
+	}
+
+	return []string{}, nil
+}
+
+// Try to reuse a single targets key across repositories.
+// FIXME: Unfortunately, short of forking Notary or sending a PR upstream, there isn't an easy way to prevent it
+// from automagically creating a new, local targets key per TUF metadata repository. We fix this here by undoing
+// more than one new, local targets key, and reusing any existing local targets key, just like the way Notary
+// reuses the root key.
+func reuseTargetsKey(r client.Repository) error {
+	var (
+		err                                error
+		thisTargetsKeyID, thatTargetsKeyID string
+	)
+
+	// Get all known targets keys.
+	targetsKeyList := r.GetCryptoService().ListKeys(data.CanonicalTargetsRole)
+	// Try to extract a single targets key we can reuse.
+	switch len(targetsKeyList) {
+	case 0:
+		err = fmt.Errorf("no targets key despite having initialized a repo")
+	case 1:
+		log.Debug("Nothing to do, only one targets key available")
+	case 2:
+		// First, we publish current changes to repository in order to list roles.
+		// FIXME: Find a find better way to list roles w/o publishing changes first.
```

Comment:
Why exactly do we need to publish the changes first?

Reply:
Also, could we have a bit more verbose logging here?

Reply:
Unfortunately, without first publishing, we are unable to list roles, IIRC from experience. Perhaps @justincormack could help to confirm, here?

Also, what do you mean by verbose logging? Where would you like to see them?